compiler/src/lnntoamm/Fn.ts

```diff
@@ -0,0 +1,235 @@
+  name: string | null
+  ast: LPNode
+  // the scope this function is defined in is the `par`
+  scope: Scope
+  args: Args
+  retTy: Type
+  body: Stmt[]
+  // not used by this class, but used by Statements
+  metadata: MetaData
+  fnType: FunctionType
+
+  get argNames(): string[] {
+    return Object.keys(this.args);
+  }
+
+  constructor(
+    ast: LPNode,
+    scope: Scope,
+    name: string | null,
+    args: Args,
+    retTy: Type | null,
+    body: Stmt[],
+    metadata: MetaData = null,
+  ) {
+    this.ast = ast;
+    this.scope = scope;
+    this.name = name;
+    this.args = args;
+    for (let argName of Object.keys(this.args)) {
+      if (this.args[argName].ty === null) {
+        this.args[argName].ty = Type.generate();
+      }
+    }
+    this.retTy = retTy !== null ? retTy : Type.generate();
+    this.body = body;
+    this.metadata = metadata !== null ? metadata : new MetaData(scope);
+    this.fnType = new FunctionType(
+      this.name,
+      Object.values(this.args).map(a => a.ty),
+      this.retTy
+    );
+  }
+
+  static fromFunctionsAst(
+    ast: LPNode,
+    scope: Scope,
+    // metadata: MetaData = null,
+  ): Fn {
+    // TODO: inheritance
+    let metadata = new MetaData(scope);
+
+    let work = ast;
+    const name = work.get('optname').has() ? work.get('optname').get().t : null;
+    let args: Args = {};
+    if (work.get('optargs').has('arglist')) {
+      let argAsts = [
+        work.get('optargs').get('arglist'),
+        ...work.get('optargs').get('arglist').get('cdr').getAll(),
+      ];
+      argAsts.forEach(argAst => {
+        const arg = FnArg.fromArgAst(argAst, metadata)
+        args[arg.name] = arg;
+      });
+    }
+
+    const retTy = work.get('optreturntype').has() ? work.get('optreturntype').get().get('fulltypename') : 'void';
+
+    let body = [];
+    let bodyAsts: LPNode | LPNode[] = work.get('fullfunctionbody');
+    if (bodyAsts.has('functionbody')) {
+      bodyAsts = bodyAsts.get('functionbody').get('statements').getAll().map(s => s.get('statement'));
+      bodyAsts.forEach(ast => body.push(...Stmt.fromAst(ast, metadata)));
+    } else {
+      bodyAsts = bodyAsts.get('assignfunction');
+      body = Stmt.fromAst(bodyAsts, metadata);
+      const retVal = body[body.length - 1];
+      if (!(retVal instanceof Dec)) {
+        throw new Error(`illegal function body: ${bodyAsts}`);
+      }
+      body.push(new Exit(bodyAsts, retVal.ref()));
+    }
+
+    return new Fn(
+      ast,
+      new Scope(scope),
+      name,
+      args,
+      Type.getFromTypename(retTy, scope),
+      body,
+    );
+  }
+
+  static fromFunctionbody(
+    ast: LPNode,
+    scope: Scope,
+  ): Fn {
+    let body = [];
+    let metadata = new MetaData(scope);
+    ast.get('statements').getAll().map(s => s.get('statement')).forEach(ast => body.push(...Stmt.fromAst(ast, metadata)));
+    return new Fn(
+      ast,
+      new Scope(scope),
+      null,
+      {},
+      opcodes().get('void'),
```

Comment:
Can be fixed later, but conditional function bodies could be something other than void

Reply:
I thought we're doing if statements, not if expressions?

Reply:
Oh, I mean that `if myConditional { <fnBody> }` is composed of `ifLiteral` `conditionalCheck` `fn` where the block-looking function form is allowed, but it could also be `if myConditional someFunction` or even (though silly) `if myConditional fn (): retType { ... }` or `if myConditional fn () = myRetVal`

Reply:
right - would this allow for things like this? if not, does it make sense to require that the fn returns `void` to avoid confusion?
```alan
fn add1(n: int64) = add(1, n);
let x = if y == 0 add1(1) else add1(y);
```

Reply:
Not exactly, but this:

```ln
fn myFn(x: int64): Result<int64> {
  fn add1(n: int64) = add(1, n);
  if n == 0 add1(1) else add1(x)
}
```

Would work? The return type of the inner function becomes the return type of the outer function.

Reply:
That makes it identical in behavior to:

`fn myFn(x: in64): Result<int64> {
  if x == 0 {
    return ok(2);
  } else {
    return x + 1;
  }
}
```